### DISCUSSION TITLE
Make it less complicated.

### DISCUSSION THREAD

**Discussion by danielmrussell** [PowerShell/PowerShell#19766] [Enhancements]
### Summary of the new feature / enhancement

Make upgrading to Powershell 7.x easy. It's so complicated and frustrating. There should be a simple installer that gives the option to migrate everything, including the Developer shortcuts especially, as well as $PROFILE (either by copying or symlink--give an option). It's crazy how much time it takes to transition and how error prone and brittle the process is. Developers have enough to worry about without all this craziness!

### Proposed technical implementation details (optional)

Provide a graphical installer with options that automate all of this. Make developers especially a focus.

**Comment by kasini3000** [comment]
Powershell is portable application, which means you can unzip, copy and run it. 
This is suitable for Windows, Linux, and ARM CPUs.

On win:
Iex "&{$(irm https://aka.ms/install-powershell.ps1 ）} - UseMSI - Quiet“

linux:
https://aka.ms/install-powershell.ps1

On Linux:
https://gitee.com/chuanjiao10/kasini3000_agent_linux

**Comment by 237dmitry** [comment]
>  as well as $PROFILE

In my case powershell and pwsh profiles are not full compatible.

**Comment by jhoneill** [comment]
> > as well as $PROFILE
> 
> In my case powershell and pwsh profiles are not full compatible.

This is why the profile directories are kept separate.  

I have made mine compatible on Windows PowerShell, PWSH 7 on Windows and PWSH on Linux (WSL), and created symbolic links to run the same file on all of them
For any non-trivial profile it is the **user's call**  whether they want to do extra work to keep things in sync but avoid compatibility issues,  or do extra work for compatibility but avoid sync issues.  But either way there is work.  

I think what the OP is calling for would save a small amount of setup for people like me (who I think are the minority) and break things for @237dmitry  (and I think the majority, or at least a considerable minority) _without an obvious fix_.   

Telling users "Put a profile _here_ if you're happy it works _with this version_" is OK IMHO. Changing Windows PowerShell so it goes to a different directory if there is no `\DocumentsWindowsPowerShell` and to search the `\program files\PowerShell\modules` folder will result in things which don't work down level being loaded and probably can't be done at this stage. (Windows PowerShell is not changing any functionality only fixing bugs and security issues).  Changing PowerShell 7 to load a Windows PowerShell 5 profile which might not be correct for 7 will also cause problems.  So I think what we have is the "least worst" of the available choices  

I totally agree with the sentiment of the OP. We have plenty of people who have been slow to get off Windows PowerShell and onto 6/7. Anything that can be done to make their lives easier is a plus. But something which migrates _everything_ is likely to migrate things which should not be moved. My _WindowsPowerShell_ directory has been a symbolic link to my personal one drive since it was called "Skydrive" and contains 15 year old junk that I've never bothered to clean up, so if the PowerShell 7.4 preview installer said "hey I'll copy that for you"  it would be the wrong thing.  So anything which tries to make this process easier comes with a risk of actually making matters worse.

**Comment by 237dmitry** [comment]
> I have made mine compatible on Windows PowerShell, PWSH 7 on Windows and PWSH on Linux (WSL), and created symbolic links to run the same file on all of them

It could be make a compatible profile, create symbolic links to it. But for this, it is necessary to organize branching in the profile. I think that this is superfluous, quite satisfied with the current state of affairs. My Windows PowerShell profile contains everything, and the PowerShell Core profile is rather the main function, if it could be compared it like that, and everything else is placed in external files (functions, classes, psstyle, custom formats and prompt files). As for Windows PowerShell I use it very rarely nowadays and did not edit its profile for a long time.

**Comment by kilasuit** [comment]
I do have a single profile that has certain checks depending on whether I am in ISE, VSCode, WinTerminal or a PowerShell v5.1 or 7.x session 

Unfortunately this is going to be a very much user and device specific thing that we can't make simpler than it already is without a really **really** heavy amount of investment. 

That said I am going to mark this for the engine & interactive console teams (I'm also a part of the Interactive team) so that they can also review this ask in case they have any further thoughts on your suggestion but personally I think this is very unlikely to happen anytime soon if it it even something that is likely to be possible without being too brittle and breaking people in the process. It certainly would be a nice to have if at all possible though, so thanks for raising it for consideration!

**Comment by theJasonHelmick** [comment]
The interactive working group has reviewed this issue. As you can see from the conversation - this is very complicated issue.  At the moment, there are more questions than answers about the best direction and technical implimentation. We look forward to learning from the community on concrete steps we could take to further resolve this pain.  We will move this to a discussion to gather further conversation.
